src/Path/Solver.hs

```diff
@@ -131,29 +131,28 @@ solve cs
   = fmap (map (uncurry toSolution) . IntMap.toList)
   . execState mempty
   . evalState (Seq.empty :: Seq.Seq (Caused (Equation Value)))
-  . evalState (mempty :: IntMap.IntMap (Set.Set (Caused (Equation Value))))
   $ do
-    visit cs
-    process
-  where visit cs = for_ cs each
-        each q@(t1 :===: t2 :@ c) = do
-          _S <- get
-          case () of
-            _ | Just s <- solutions _S (metaNames (fvs t1 <> fvs t2)) -> simplify (apply s q) >>= visit
-              | Just (m, sp) <- pattern t1 -> solve (m := abstractLam sp t2 :@ c)
-              | Just (m, sp) <- pattern t2 -> solve (m := abstractLam sp t1 :@ c)
-              | otherwise -> enqueue q
-
-        process = do
+    stuck <- fmap fold . execState (mempty :: IntMap.IntMap (Set.Set (Caused (Equation Value)))) $ do
+      modify (flip (foldl' (Seq.|>)) cs)
+      step
+    unless (Prelude.null stuck) $ for_ stuck throwMismatch -- FIXME: throw a single error comprised of all of them
```

Comment:
I’d like to throw the entire set of unsolvable constraints, but haven’t got that far yet.

Likewise, I’d like to move the reporting of type mismatch errors out of `simplify` and instead report failure there by a `Nothing` value + deactivating the stuck constraint; we’ll do more work in erroring cases, but we’ll be able to report all failures. (Enqueueing would also have to be changed to be able to record stuck constraints that aren’t dependent on any particular metavariable.)